Library/OcCryptoLib/BigNumPrimitives.c

```diff
@@ -0,0 +1,1010 @@
+  //
+  ASSERT (NumWords > 0);
+
+  for (Index = (A->NumWords - 1); Index > 0; --Index) {
+    Result->Words[Index + NumWords] = (A->Words[Index] << NumBits) | (A->Words[Index - 1] >> (OC_BN_WORD_NUM_BITS - NumBits));
+  }
+  //
+  // Handle the edge-cases at the beginning and the end of the value.
+  //
+  Result->Words[NumWords] = A->Words[0] << NumBits;
+  Result->Words[A->NumWords + NumWords] = A->Words[A->NumWords - 1] >> (OC_BN_WORD_NUM_BITS - NumBits);
+  //
+  // Zero everything outside of the previously set ranges.
+  //
+  ZeroMem (&Result->Words[A->NumWords + NumWords + 1], (Result->NumWords - NumWords - A->NumWords - 1) * OC_BN_WORD_SIZE);
+  ZeroMem (Result->Words, NumWords * OC_BN_WORD_SIZE);
+}
+
+/**
+  Shifts A left by Exponent Bits for 0 < Exponent < #Bits(Word).
+
+  @param[in,out] A         The base.
+  @param[in]     Exponent  The Bit shift exponent.
+
+**/
+STATIC
+VOID
+BigNumLeftShiftBitsSmall (
+  IN OUT OC_BN  *A,
+  IN     UINT8  Exponent
+  )
+{
+  UINTN Index;
+
+  ASSERT (A != NULL);
+  //
+  // Exponent must not be 0 because a shift of a Word by its Bit width or
+  // larger is Undefined Behaviour.
+  //
+  ASSERT (Exponent > 0);
+  ASSERT (Exponent < OC_BN_WORD_NUM_BITS);
+
+  for (Index = (A->NumWords - 1); Index > 0; --Index) {
+    A->Words[Index] = (A->Words[Index] << Exponent) | (A->Words[Index - 1] >> (OC_BN_WORD_NUM_BITS - Exponent));
+  }
+  A->Words[0] <<= Exponent;
+}
+
+/**
+  Shifts A right by Exponent Bits for 0 < Exponent < #Bits(Word).
+
+  @param[in,out] A         The base.
+  @param[in]     Exponent  The Bit shift exponent.
+
+**/
+STATIC
+VOID
+BigNumRightShiftBitsSmall (
+  IN OUT OC_BN  *A,
+  IN     UINT8  Exponent
+  )
+{
+  UINTN Index;
+
+  ASSERT (A != NULL);
+  //
+  // Exponent must not be 0 because a shift of a Word by its Bit width or
+  // larger is Undefined Behaviour.
+  //
+  ASSERT (Exponent > 0);
+  ASSERT (Exponent < OC_BN_WORD_NUM_BITS);
+
+  for (Index = 0; Index < (A->NumWords - 1); ++Index) {
+    A->Words[Index] = (A->Words[Index] >> Exponent) | (A->Words[Index + 1] << (OC_BN_WORD_NUM_BITS - Exponent));
+  }
+  A->Words[Index] >>= Exponent;
+}
+
+#if defined(_MSC_VER) && !defined(__clang__)
+  #include <intrin.h>
+  #pragma intrinsic(_umul128)
+#endif
+
+/**
+  Calculates the product of A and B.
+
+  @param[out] Hi  Buffer in which the high Word of the result is returned.
+  @param[in]  A   The multiplicant.
+  @param[in]  B   The multiplier.
+
+  @returns  The low Word of the result.
+
+**/
+OC_BN_WORD
+BigNumWordMul (
+  OUT OC_BN_WORD  *Hi,
+  IN  OC_BN_WORD  A,
+  IN  OC_BN_WORD  B
+  )
+{
+  ASSERT (Hi != NULL);
+
+  if (OC_BN_WORD_SIZE == 4) {
+    UINT64 Result = (UINT64)A * B;
+    //
+    // FIXME: The subtraction in the shift is a dirty hack to shut up MSVC.
+    //
+    *Hi = (OC_BN_WORD)(Result >> (OC_BN_WORD_NUM_BITS - (OC_BN_WORD_SIZE != 4)));
+    return (OC_BN_WORD)Result;
+  } else if (OC_BN_WORD_SIZE == 8) {
+  #if !defined(_MSC_VER) || defined(__clang__)
```

Comment:
I think we discussed this once that it should be guarded by MDE_CPU_X64 (or something similar) as it will not compile on 32-bit machines. Once it is not set, the slow software version (that is now commented out) should be used.

Reply:
After spending a while on fixing type warnings, I confirmed this compiles with both MSVC and Clang for the IA32 target just fine